File: scripts/ingest.ts
import { datetime } from 'next/dist/compiled/@edge-runtime/primitives';

interface RawDocument {
  id: string;
  court: string;
  judge: string;
  document_type: string;
  content_format: 'scanned_pdf' | 'text';
  content_payload: string;
  date_filed: string;
}

interface NormalizedRecord {
  id: string;
  judge_name: string;
  court: string;
  charge_type: string;
  sentence_months: number;
  guideline_range: string;
  deviation_months: number;
  ingested_at: string;
}

class IngestionPipeline {
  private rawBucketS3 = 's3://benchmark-justice-raw-ingest-bucket/';

  constructor() {
    console.log('[IngestionPipeline] Initializing Ingestion & OCR pipeline standard engines...');
  }

  fetchSourceFeed(sourceName: string): RawDocument | null {
    console.log(`[IngestionPipeline] Connecting to raw API feed: ${sourceName}`);
    if (sourceName === 'CourtListener') {
      return {
        id: 'opin-104921',
        court: 'U.S. District Court, Southern District of Florida',
        judge: 'Aileen Cannon',
        document_type: 'opinion',
        content_format: 'scanned_pdf',
        content_payload: '[Scanned Text Overlay - Defendant found guilty of possessing controlled substances. Sentenced to 120 months imprisonment. Guidelines variance evaluated.]',
        date_filed: '2024-03-12'
      };
    }
    return null;
  }

  ocrLayer(rawDoc: RawDocument): string {
    if (rawDoc.content_format === 'scanned_pdf') {
      console.log(`[IngestionPipeline] Triggering OCR Layer (TesseractEngine) on document ${rawDoc.id}`);
      const textExtracted = rawDoc.content_payload
        .replace('[Scanned Text Overlay - ', '')
        .replace(']', '');
      console.log('[IngestionPipeline] OCR Extraction completed successfully.');
      return textExtracted;
    }
    return rawDoc.content_payload;
  }

  validateSchema(normalizedItem: NormalizedRecord, entityType: string): boolean {
    console.log(`[IngestionPipeline] Validating database schema constraints for ${entityType}...`);
    const requiredKeys: (keyof NormalizedRecord)[] = ['id', 'judge_name', 'court', 'sentence_months'];
    const valid = requiredKeys.every(k => normalizedItem[k] !== undefined);
    console.log(`[IngestionPipeline] Schema validation check: ${valid ? 'PASSED' : 'FAILED'}`);
    return valid;
  }

  standardizeMapping(rawText: string): NormalizedRecord {
    console.log('[IngestionPipeline] Mapping unstructured fields to standardized schema columns...');
    const chargeType = rawText.toLowerCase().includes('controlled substances') ? 'Drug Offenses' : 'General';
    const sentenceMonths = rawText.toLowerCase().includes('120 months') ? 120 : 0;

    return {
      id: 'cannon',
      judge_name: 'Aileen Cannon',
      court: 'U.S. District Court, Southern District of Florida',
      charge_type: chargeType,
      sentence_months: sentenceMonths,
      guideline_range: '108-135 months',
      deviation_months: 0,
      ingested_at: new Date().toISOString()
    };
  }

  uploadToRawS3(recordId: string, payload: any): void {
    const s3Path = `${this.rawBucketS3}${recordId}.json`;
    console.log(`[IngestionPipeline] Archiving raw data to AWS S3: ${s3Path}`);
    console.log(`[IngestionPipeline] Successfully archived record ${recordId} to raw lake.`);
  }

  runPipeline(): void {
    console.log('--- STARTING DAILY BATCH INGESTION JOB ---');

    // 1. Fetch
    const rawDoc = this.fetchSourceFeed('CourtListener');
    if (!rawDoc) return;

    // 2. Archive
    this.uploadToRawS3(rawDoc.id, rawDoc);

    // 3. OCR
    const extractedText = this.ocrLayer(rawDoc);

    // 4. Map
    const normalized = this.standardizeMapping(extractedText);

    // 5. Validate
    if (this.validateSchema(normalized, 'judge')) {
      console.log('[IngestionPipeline] Pushing normalized record to PostgreSQL & ElasticSearch queues.');
      console.log(JSON.stringify(normalized, null, 2));
    }

    console.log('--- INGESTION JOB COMPLETED SUCCESSFULLY ---');
  }
}

// Execute
const pipeline = new IngestionPipeline();
pipeline.runPipeline();

File: scripts/ingest.py
#!/usr/bin/env python3
"""
Benchmark Justice™ Data Ingestion & OCR Pipeline Simulation
Represents Phase 1-4 data pipeline executing OCR, schemas validation, standardization, and raw storage.
"""

import os
import sys
import json
import logging
from datetime import datetime

# Setup logger
logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s] %(message)s")
logger = logging.getLogger("IngestionPipeline")

class IngestionPipeline:
    def __init__(self):
        self.raw_bucket_s3 = "s3://benchmark-justice-raw-ingest-bucket/"
        logger.info("Initializing Ingestion & OCR pipeline standard engines...")

    def fetch_source_feed(self, source_name: str) -> dict:
        """Simulates fetching source documents/JSON records from legal feeds."""
        logger.info(f"Connecting to raw API feed: {source_name}")
        if source_name == "CourtListener":
            return {
                "id": "opin-104921",
                "court": "U.S. District Court, Southern District of Florida",
                "judge": "Aileen Cannon",
                "document_type": "opinion",
                "content_format": "scanned_pdf",
                "content_payload": "[Scanned Text Overlay - Defendant found guilty of possessing controlled substances. Sentenced to 120 months imprisonment. Guidelines variance evaluated.]",
                "date_filed": "2024-03-12"
            }
        elif source_name == "Congress.gov":
            return {
                "id": "bill-hr-104",
                "sponsor": "Chuck Schumer",
                "title": "A Bill for Judicial Reform",
                "vote_rolls": [
                    {"legislator_id": "schumer", "vote": "yea"},
                    {"legislator_id": "cruz", "vote": "nay"}
                ],
                "date_voted": "2024-04-01"
            }
        return {}

    def ocr_layer(self, raw_doc: dict) -> str:
        """Runs simulated OCR on scanned PDFs or returns text directly for raw payload."""
        if raw_doc.get("content_format") == "scanned_pdf":
            logger.info(f"Triggering OCR Layer (TesseractEngine) on document {raw_doc['id']}")
            # Simulated OCR extraction
            text_extracted = raw_doc["content_payload"].replace("[Scanned Text Overlay - ", "").replace("]", "")
            logger.info("OCR Extraction completed successfully.")
            return text_extracted
        return raw_doc.get("content_payload", "")

    def validate_schema(self, normalized_item: dict, entity_type: str) -> bool:
        """Checks schema integrity and constraints validation."""
        logger.info(f"Validating database schema constraints for {entity_type}...")
        if entity_type == "judge":
            required_keys = ["id", "judge_name", "court", "sentence_months"]
            valid = all(k in normalized_item for k in required_keys)
            logger.info(f"Schema validation check: {'PASSED' if valid else 'FAILED'}")
            return valid
        return True

    def standardize_mapping(self, raw_text: str) -> dict:
        """Maps unstructured document details into normalized database values."""
        logger.info("Mapping unstructured fields to standardized schema columns...")
        # Simulated NLP matching engine
        charge_type = "Drug Offenses" if "controlled substances" in raw_text.lower() else "General"
        sentence_months = 120 if "120 months" in raw_text.lower() else 0
        
        normalized = {
            "id": "cannon",
            "judge_name": "Aileen Cannon",
            "court": "U.S. District Court, Southern District of Florida",
            "charge_type": charge_type,
            "sentence_months": sentence_months,
            "guideline_range": "108-135 months",
            "deviation_months": 0, # Right at the median guidelines
            "ingested_at": datetime.utcnow().isoformat()
        }
        return normalized

    def upload_to_raw_s3(self, record_id: str, payload: dict):
        """Simulates raw storage archive in AWS S3."""
        s3_path = f"{self.raw_bucket_s3}{record_id}.json"
        logger.info(f"Archiving raw data to AWS S3: {s3_path}")
        # In production, this would call boto3.client('s3').put_object()
        logger.info(f"Successfully archived record {record_id} to raw lake.")

    def run_pipeline(self):
        logger.info("--- STARTING DAILY BATCH INGESTION JOB ---")
        
        # 1. Fetch
        raw_doc = self.fetch_source_feed("CourtListener")
        
        # 2. Archive Raw
        self.upload_to_raw_s3(raw_doc["id"], raw_doc)
        
        # 3. OCR Layer
        extracted_text = self.ocr_layer(raw_doc)
        
        # 4. Standardize Mapping
        normalized = self.standardize_mapping(extracted_text)
        
        # 5. Validate Schema
        if self.validate_schema(normalized, "judge"):
            logger.info("Pushing normalized record to PostgreSQL & ElasticSearch queues.")
            logger.info(json.dumps(normalized, indent=2))
        
        logger.info("--- INGESTION JOB COMPLETED SUCCESSFULLY ---")

if __name__ == "__main__":
    pipeline = IngestionPipeline()
    pipeline.run_pipeline()
